### ISSUE TITLE
Add hook for measuring components

### ISSUE THREAD

**Issue by kripod** [kripod/react-hooks#19]
## Motivation

One may need to know the exact size of a DOM node, to know whether it's big or small enough.

## Details

[How can I measure a DOM node? – React Hooks FAQ](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node):

```js
function MeasureExample() {
  const [rect, ref] = useClientRect();
  return (
    <>
      <h1 ref={ref}>Hello, world</h1>
      {rect !== null &&
        <h2>The above header is {Math.round(rect.height)}px tall</h2>
      }
    </>
  );
}

function useClientRect() {
  const [rect, setRect] = useState(null);
  const ref = useCallback(node => {
    if (node !== null) {
      setRect(node.getBoundingClientRect());
    }
  }, []);
  return [rect, ref];
}
```

**Comment by cmoog** [comment]
Seems like you'd need to be able to refresh the measurement, maybe on a scroll event or after some animation etc. What do you think about the following refactor?

**Simple usage**
```jsx
const [rect, ref, refresh] = useNodeRect()
```

**Implementation**
```jsx
function useNodeRect() {
  const [rect, setRect] = useState(null)
  const savedRef = useRef()

  const ref = useCallback(node => {
    if (node.getBoundingClientRect) {
      setRect(node.getBoundingClientRect())
      savedRef.current = node
    } else if (savedRef.current) {
      setRect(savedRef.current.getBoundingClientRect())
    }
  }, [])
  const refresh = ref

  return [rect, ref, refresh]
}

```

**Comment by kripod** [comment]
Personally, I'm thinking about using the `ResizeObserver` API, but it comes with the cost of polyfilling as of today:

```ts
import { useCallback, useMemo, useState, useEffect } from 'react';
import ResizeObserver from 'resize-observer-polyfill';

export default function useSize() {
  const [size, setSize] = useState<Readonly<[number, number]>>([0, 0]);
  const ro = useMemo(
    () =>
      new ResizeObserver(([entry]) => {
        setSize([entry.contentRect.width, entry.contentRect.height]);
      }),
    [],
  );

  const ref = useCallback(
    node => {
      if (node) {
        ro.observe(node);
      }
    },
    [ro],
  );

  useEffect(() => {
    return () => {
      ro.disconnect();
    };
  }, [ro]);

  return [size, ref];
}
```

Also, I'm not entirely sure about the `ro.disconnect()` part.

**Comment by cmoog** [comment]
Looks like that's what `react-use` has as well.

**Comment by kripod** [comment]
Here is an improved snippet of code, which makes sure not to observe the same target multiple times:

```ts
import { useCallback, useMemo, useState } from 'react';
import ResizeObserver from 'resize-observer-polyfill';

export default function useSize(): [
  Readonly<[number, number]>,
  (element: HTMLElement) => void,
] {
  const [size, setSize] = useState<Readonly<[number, number]>>([0, 0]);

  const ro = useMemo(
    () =>
      new ResizeObserver(([entry]) => {
        setSize([entry.contentRect.width, entry.contentRect.height]);
      }),
    [],
  );

  const ref = useCallback(
    (element: HTMLElement) => {
      // Avoid observing the same target multiple times
      ro.disconnect();

      if (element) {
        ro.observe(element);
      }
    },
    [ro],
  );

  return [size, ref];
}
```

**Comment by kripod** [comment]
This has been resolved by bb74f0bbd6404e7d654f62f2b887403ccaf16afa.
